### PR TITLE
KEP-4222: Support CBOR encoding for apply requests.

### PR DESCRIPTION
**Pull request by benluddy**
#### What type of PR is this?

/kind cleanup
/sig api-machinery

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://kep.k8s.io/4222
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go [hunk 1]
```diff
@@ -634,7 +645,15 @@ func (p *patcher) patchResource(ctx context.Context, scope *RequestScope) (runti
 			fieldManager:       scope.FieldManager,
 		}
 	// this case is unreachable if ServerSideApply is not enabled because we will have already rejected the content type
-	case types.ApplyPatchType:
+	case types.ApplyPatchType, "application/apply-patch+cbor":
+		unmarshalFn, unmarshalStrictFn := yaml.Unmarshal, yaml.UnmarshalStrict
+		if p.patchType == "application/apply-patch+cbor" {
```
Comment (jpbetz) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go:650]: Worth adding a const for "application/apply-patch+cbor" ?

File: staging/src/k8s.io/apiserver/pkg/endpoints/installer.go [hunk 2]
```diff
@@ -877,6 +877,9 @@ func (a *APIInstaller) registerResourceHandlers(path string, storage rest.Storag
 				string(types.StrategicMergePatchType),
 				string(types.ApplyPatchType),
 			}
+			if utilfeature.TestOnlyFeatureGate.Enabled(features.TestOnlyCBORServingAndStorage) {
+				supportedTypes = append(supportedTypes, "application/apply-patch+cbor")
```
Comment (jpbetz) [inline staging/src/k8s.io/apiserver/pkg/endpoints/installer.go:881]: Introduce const for "application/apply-patch+cbor"?

**Comment by benluddy** [comment]
```
=== FAIL: k8s.io/client-go/dynamic TestGoldenRequest/apply (0.00s)
    golden_test.go:211: unexpected difference from expected bytes:
          bytes.Join({
          	"PATCH /apis/flops/v1alpha1/namespaces/mops/flips/mips/fin?force=",
          	"true HTTP/1.1\r\nHost: example.com\r\nAccept: application/json\r\nAcce",
          	"pt-Encoding: gzip\r\nContent-Length: 2",
        - 	"9",
        + 	"8",
          	"\r\nContent-Type: application/apply-patch+yaml\r\nUser-Agent: TestGo",
          	"ldenRequest\r\n\r\n{\"metadata\":{\"name\":\"mips\"}}",
        - 	"\n",
          }, "")

=== FAIL: k8s.io/client-go/dynamic TestGoldenRequest/applystatus (0.00s)
    golden_test.go:211: unexpected difference from expected bytes:
          bytes.Join({
          	"PATCH /apis/flops/v1alpha1/namespaces/mops/flips/mips/status?for",
          	"ce=true HTTP/1.1\r\nHost: example.com\r\nAccept: application/json\r\nA",
          	"ccept-Encoding: gzip\r\nContent-Length: 2",
        - 	"9",
        + 	"8",
          	"\r\nContent-Type: application/apply-patch+yaml\r\nUser-Agent: TestGo",
          	"ldenRequest\r\n\r\n{\"metadata\":{\"name\":\"mips\"}}",
        - 	"\n",
          }, "")
```

This is happening because this PR unifies the serialization of apply configurations into request bodies between the dynamic and typed clients. The dynamic client currently accepts an `*Unstructured` as the apply configuration and encodes it with the `UnstructuredJSONScheme`, which ultimately calls `json.NewEncoder(...).Encode(...)` and appends a newline: https://github.com/kubernetes-sigs/json/blob/cfa47c3a1cc8ff0eff148aa9ec5b0226d0909e87/internal/golang/encoding/json/stream.go#L214-L220. The typed clients call `json.Marshal`, which doesn't.

Is there an argument for preserving the trailing newline in the dynamic client (or adding it to requests from typed clients)?

File: staging/src/k8s.io/client-go/util/apply/apply.go [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+	"fmt"
+
+	cbor "k8s.io/apimachinery/pkg/runtime/serializer/cbor/direct"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/json"
+	"k8s.io/client-go/features"
+	"k8s.io/client-go/rest"
+)
+
+func NewRequest(client rest.Interface, c interface{}) (*rest.Request, error) {
```
Comment (jpbetz) [inline staging/src/k8s.io/client-go/util/apply/apply.go:29]: +godoc?

**Review by jpbetz — COMMENTED** [review]
/lgtm

One minor non-blocking comment.

@deads2k do you have any concerns? I'm comfortable seeing this merge.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go [hunk 4]
```diff
@@ -129,10 +130,18 @@ func PatchResource(r rest.Patcher, scope *RequestScope, admit admission.Interfac
...
 
-		baseContentType := runtime.ContentTypeJSON
-		if patchType == types.ApplyPatchType {
+		var baseContentType string
+		switch patchType {
+		case types.ApplyYAMLPatchType:
 			baseContentType = runtime.ContentTypeYAML
+		case types.ApplyCBORPatchType:
+			// The allowlist check (above) should prevent CBOR requests from reaching
+			// this point without enabling the right feature gate.
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go:139]: even so, gate please.

File: staging/src/k8s.io/client-go/dynamic/simple.go [hunk 5]
```diff
@@ -355,21 +351,16 @@ func (c *dynamicResourceClient) Apply(ctx context.Context, name string, obj *uns
 	}
 	patchOpts := opts.ToPatchOptions()
 
-	result := c.client.client.
-		Patch(types.ApplyPatchType).
```
Comment (deads2k) [inline staging/src/k8s.io/client-go/dynamic/simple.go:359]: note to self, find where this is now being set.

File: staging/src/k8s.io/client-go/util/apply/apply.go [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+	pt := types.ApplyYAMLPatchType
+	marshal := json.Marshal
+
+	if features.TestOnlyFeatureGates.Enabled(features.TestOnlyClientAllowsCBOR) && features.TestOnlyFeatureGates.Enabled(features.TestOnlyClientPrefersCBOR) {
```
Comment (deads2k) [inline staging/src/k8s.io/client-go/util/apply/apply.go:33]: We'll want to do better for a beta.

**Comment by deads2k** [comment]
I'm ok with newlines changing.

File: staging/src/k8s.io/client-go/util/apply/apply.go (hunk 3 above)
Comment (benluddy) [reply staging/src/k8s.io/client-go/util/apply/apply.go]: /hold cancel

**Pull request merged**